crypto/txsigner_eip155.go

```diff
@@ -2,90 +2,166 @@ package crypto
+
+	arenaPool.Put(RLP)
+
+	return types.BytesToHash(hash)
+}
+
+// Sender returns the sender of the transaction
+func (signer *EIP155Signer) Sender(tx *types.Transaction) (types.Address, error) {
+	if tx.ChainID() == nil || tx.ChainID().Cmp(big.NewInt(0)) == 0 {
+		return signer.HomesteadSigner.Sender(tx)
 	}
 
-	if !protected {
-		return (&FrontierSigner{}).Sender(tx)
+	v, r, s := tx.RawSignatureValues()
+
+	// Checking one of the values is enought since they are inseparable
+	if v == nil {
+		return types.Address{}, errors.New("Sender method: Unknown signature")
 	}
 
-	// Reverse the V calculation to find the original V in the range [0, 1]
-	// v = CHAIN_ID * 2 + 35 + {0, 1}
-	mulOperand := big.NewInt(0).Mul(big.NewInt(int64(e.chainID)), big.NewInt(2))
-	bigV.Sub(bigV, mulOperand)
-	bigV.Sub(bigV, big35)
+	// Reverse the V calculation to find the parity of the Y coordinate
+	// v = CHAIN_ID * 2 + 35 + {0, 1} -> {0, 1} = v - 35 - CHAIN_ID * 2
+
+	a := big.NewInt(0)
+	b := big.NewInt(0)
+	parity := big.NewInt(0)
+
+	// a = v - 35
+	a.Sub(v, big35)
+
+	// b = CHAIN_ID * 2
+	b.Mul(big.NewInt(int64(signer.chainID)), big.NewInt(2))
 
-	return recoverAddress(e.Hash(tx), r, s, bigV, e.isHomestead)
+	// parity = a - b
+	parity.Sub(a, b)
+
+	return recoverAddress(signer.Hash(tx), r, s, parity, true)
 }
 
-// SignTx signs the transaction using the passed in private key
-func (e *EIP155Signer) SignTx(
-	tx *types.Transaction,
-	privateKey *ecdsa.PrivateKey,
-) (*types.Transaction, error) {
+// SingTx takes the original transaction as input and returns its signed version
+func (signer *EIP155Signer) SignTx(tx *types.Transaction, privateKey *ecdsa.PrivateKey) (*types.Transaction, error) {
+	if tx.ChainID() == nil || tx.ChainID().Cmp(big.NewInt(0)) == 0 {
+		return signer.HomesteadSigner.SignTx(tx, privateKey)
+	}
+
 	tx = tx.Copy()
 
-	h := e.Hash(tx)
+	hash := signer.Hash(tx)
 
-	sig, err := Sign(privateKey, h[:])
+	signature, err := Sign(privateKey, hash[:])
 	if err != nil {
 		return nil, err
 	}
 
-	r := new(big.Int).SetBytes(sig[:32])
-	s := new(big.Int).SetBytes(sig[32:64])
-	v := new(big.Int).SetBytes(e.calculateV(sig[64]))
+	r := new(big.Int).SetBytes(signature[:32])
+	s := new(big.Int).SetBytes(signature[32:64])
+
+	if s.Cmp(secp256k1NHalf) > 0 {
+		return nil, errors.New("SignTx method: S must be inclusively lower than secp256k1n/2")
+	}
+
+	v := new(big.Int).SetBytes(signer.calculateV(signature[64]))
 
 	tx.SetSignatureValues(v, r, s)
 
 	return tx, nil
 }
 
-// calculateV returns the V value for transaction signatures. Based on EIP155
-func (e *EIP155Signer) calculateV(parity byte) []byte {
-	reference := big.NewInt(int64(parity))
-	reference.Add(reference, big35)
+// Private method calculateV returns the V value for the EIP-155 transactions
+//
+// V is calculated by the formula: {0, 1} + CHAIN_ID * 2 + 35 where {0, 1} denotes the parity of the Y coordinate
+func (signer *EIP155Signer) calculateV(parity byte) []byte {
+	a := big.NewInt(0)
```

Comment:
Please, return this function as it was. It just adds to the complexity of the code.